data-upgrade-news/src/main/java/org/exoplatform/news/upgrade/jcr/NewsArticlesViewsCountUpgrade.java

```diff
@@ -0,0 +1,102 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.exoplatform.news.upgrade.jcr;
+
+import org.exoplatform.commons.upgrade.UpgradeProductPlugin;
+import org.exoplatform.container.xml.InitParams;
+import org.exoplatform.services.cms.BasePath;
+import org.exoplatform.services.jcr.RepositoryService;
+import org.exoplatform.services.jcr.ext.app.SessionProviderService;
+import org.exoplatform.services.jcr.ext.common.SessionProvider;
+import org.exoplatform.services.jcr.ext.hierarchy.NodeHierarchyCreator;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+
+import javax.jcr.Node;
+import javax.jcr.NodeIterator;
+import javax.jcr.Session;
+
+public class NewsArticlesViewsCountUpgrade extends UpgradeProductPlugin {
+
+    private static final Log LOG                           = ExoLogger.getLogger(NewsArticlesViewsCountUpgrade.class.getName());
+
+    private NodeHierarchyCreator nodeHierarchyCreator;
+    private RepositoryService repositoryService;
+    private SessionProviderService sessionProviderService;
+
+    private int updatedNodes                              = 0 ;
+    public static final String NEWS_NODES_FOLDER          = "News";
+
+
+    public NewsArticlesViewsCountUpgrade(InitParams initParams, NodeHierarchyCreator nodeHierarchyCreator,
+                                         RepositoryService repositoryService,
+                                         SessionProviderService sessionProviderService) {
+        super(initParams);
+        this.nodeHierarchyCreator = nodeHierarchyCreator;
+        this.repositoryService = repositoryService;
+        this.sessionProviderService = sessionProviderService;
+    }
+
+    public int getUpdatedNodes() {
+        return updatedNodes;
+    }
+
+    @Override
+    public void processUpgrade(String oldVersion, String newVersion) {
+        long startupTime = System.currentTimeMillis();
+        LOG.info("Start upgrade of news articles viewsCount");
+        SessionProvider sessionProvider = null;
+        try {
+            sessionProvider = sessionProviderService.getSystemSessionProvider(null);
+            Session session = sessionProvider.getSession(
+                    repositoryService.getCurrentRepository()
+                            .getConfiguration()
+                            .getDefaultWorkspaceName(),
+                    repositoryService.getCurrentRepository());
+            String spacesNodePath = nodeHierarchyCreator.getJcrPath(BasePath.CMS_GROUPS_PATH) + "/spaces";
+            Node spacesRootNode = (Node) session.getItem(spacesNodePath);
+            NodeIterator iter = spacesRootNode.getNodes();
+            while (iter.hasNext()) {
```

Comment:
Could you reverse the fetching and do it through the News API with pagination ?

Reply:
I used a query to avoid looping in all spaces, Could you check ?